### DISCUSSION TITLE
[Chapter 05] 메서드 체인은 응집도를 낮출 수 있어 좋지 않다?

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/is-my-code-that-strange#40] [Chapter 04-05 (불변, 응집도)]
Page 81-82. 에는 아래와 같은 글귀가 있습니다.
> 메서드 체인은 응집도를 낮출 수 있어 좋지 않은 작성 방법입니다. 

그런데 Java 8 이상부터는 Stream API 를 많이 활용하고 있는데요.
그럼 Stream API 에 사용되는 메서드 체인도 응집도를 낮추는 좋지 않은 방법일까요?

**Comment by corock** [comment]
```java
public class Game {

    private Party party;

    /**
     * 갑옷을 장착한다.
     *
     * @param memberId 장비를 변경하고 싶은 멤버의 ID
     * @param newArmor 입을 갑옷
     */
    public void equipArmor(int memberId, Armor newArmor) {
        if (party.members[memberId].equipments.canChange) {
            party.members[memberId].equipments.armor = newArmor;
        }
    }

}
```

> 게임에서 멤버의 갑옷을 변경하는 메서드입니다.
> (...)
> 현재 예에서는 메서드 체인을 사용해서, 클래스의 꽤 깊은 곳에 있는 요소에 접근하고 있습니다.

저자가 위 문장을 언급한 의도로 저는 '속성과 오퍼레이션을 적절한 곳에 위치시키면 응집도가 높아진다.' 라는 뜻으로 이해했습니다.
이 문제를 해결하면 주로 메서드 체인 형태보다는 `foo.bar()` 처럼 하나의 .(_dot_) 모양 꼴로 나오기 때문인데요.
그럼 '스트림 API 는 메서드 체이닝으로 사용하기 때문에 응집도가 낮아지는 거 아니냐?' 라고 반문하면 이게 응집도와 관련이 있다고 봐야 하는 건가 싶습니다.

오히려 스트림 API 에 대한 통념은 응집도보단 결합도에 문제가 되는가? 로 [디미터 법칙(_Law of Demeter_)](https://tecoble.techcourse.co.kr/post/2020-06-02-law-of-demeter) 이라는 개념과 결부 지어서 언급합니다.
_Practical Object-Ortiented Design_ 라는 저서에서는 다음처럼 디미터 법칙을 언급하고 있습니다.

> use only one dot (오직 하나의 도트만을 사용하라)

책 전반에 걸쳐 프로그래밍 언어 상관없이 개념 전달에 초점을 맞추기 위해서인지, 아니면 간단한 예시를 들기 위해서인지
저자는 필드에 무지성 접근하는 모습을 보이긴 하지만 다음처럼 비교적 익숙한 코드로 수정한다 해도 응집도가 해결되는 건 아닙니다.

```java
        if (party.getMembers().getMemberBy(memberId).getEquipments().getCanChange()) {
            // (...)
        }
```

제가 알기로 위와 같은 형태는 클린 코드라는 도서에서 기차 충돌(_Train Wrek_)이라는 용어로 소개됩니다.
객체지향이라는 책에서 우문현답을 아래와 같이 설명하고 있어 공유해 드립니다.

> `filter`, `distinct` 등의 메서드는 모두 동일한 스트림 클래스의 인스턴스를 반환한다.
> (...)
> 디미터 법칙은 결합도와 관련된 것이며, 이 결합도가 문제가 되는 것은 객체의 내부 구조가 외부로 노출되는 경우로 한정된다.
>
